### ISSUE TITLE
cannot exit recursive infinite loop in tfrecord_loader.py

### ISSUE THREAD

**Issue by bvelker** [kingoflolz/mesh-transformer-jax#129]
The network loads just fine but gets stuck after printing "compiling train fn"
specifically at 
train_dataset.get_samples()

get_samples() is a (potentially)recursive function in tfrecord_loader.py which appears to train the 1st step

My question is: what could cause this to loop forever?
I've double checked the paths so that's probably not the issue (I hope not)

train.index and val.index files are formatted like your examples (with the train/val tfrecord datasets in their own respective folders)

Total (2048 token) training sequences = 1505
Total (2048 token) validation sequences = 217

Here is my config file ¯\\_(ツ)_/¯

{
  "layers": 28,
  "d_model": 4096,
  "n_heads": 16,
  "n_vocab": 50400,
  "norm": "layernorm",
  "pe": "rotary",
  "pe_rotary_dims": 64,

  "seq": 2048,
  "cores_per_replica": 8,
  "per_replica_batch": 1,
  "gradient_accumulation_steps": 16,

  "warmup_steps": 4,
  "anneal_steps": 50,
  "lr": 1.2e-5,
  "end_lr": 1.2e-6,
  "weight_decay": 0.1,
  "total_steps": 54,

  "tpu_size": 8,

  "bucket": "gptj-testing",
  "model_dir": "newModel",

  "train_set": "housefile_content_emails-subjectlines.train.index",
  "val_set": {
    "emails": "housefile_content_emails-subjectlines.val.index"
  },
  
  "eval_harness_tasks": [
  ],

  "val_batches": 1722,
  "val_every": 500,
  "ckpt_every": 500,
  "keep_every": 1500,

  "name": "1st_run",
  "wandb_project": "large_housefile_content_emails-subjectlines",
  "comment": "please work!"
}

I've been stuck on permutations of this problem for a week now. Any help would be greatly appreciated

**Comment by cirecrux** [comment]
`get_samples()` in `tfrecord_loader.py` is also looping forever for me, although it happens after `compiling train fn` during the actual fine tuning process, and I'm getting a **RecursionError** (stack overflow) instead of an infinite loop. 

I've been going through the code, and my best guess from previous versions and the most recent documentation of tfrecord_loader.py (see #16) --- is that the list `self.used` in `tfrecord_loader.py` is filling up and causing the problem, making it loop forever. 

For my own issue, I have a theory that the function `sample_once(self)` somehow isn't yielding anything to `self.sample_fn`, making it send the **StopIteration** thing recursively. But I can't figure out why or how that is (not a surprise since I just started learning coding in general). 

Sorry that I couldn't find a solution, but hopefully this will give you a start.

**Comment by cirecrux** [comment]
Ok, I figured out my own problem out: I had an empty index file (for the val.json) that I left in my config file which caused `tfrecord_loader.py`'s `get_samples()` to go into a **RecursionError** situation. 

I'm still not sure what is causing your error, but if there isn't a error message being explicitly printed out, it could just be that your dataset is just taking forever to finish compiling since it is gigantic. 

Also, I just noticed this, but your config file's `total_steps`, `warmup_steps`, and `anneal_steps` doesn't seem to be in sync with the tfrecords you have in your index file --- there should be a total of 1505 sequences in your training step, which when dividing by the gradient_accumulation_steps, there should be about 95 epochs (total steps), not 54. Your `val_batches` property of 1722 seems kind of large as well.
